### ISSUE TITLE
[Bug] MiniMax-M3 W4A16 (compressed-tensors) on 2x DGX Spark (sm_121, TP=2): serves but every token is id 0 — all-NUL output on the Triton MiniMaxSparse path; same weights correct on vLLM

### ISSUE THREAD

**Issue by robertforbes68** [sgl-project/sglang#38143]
### Checklist

- [x] I searched related issues but found no solution.
- [ ] The bug persists in the latest version. (Observed on v0.5.18; not yet re-run on v0.5.19 — the v0.5.19 notes contain no MiniMax-M3 text-model or SM121 sparse-attention changes, so I expect it to persist. Happy to re-test on request.)
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion.
- [x] Please use English.

### Describe the bug

**MiniMax-M3 (W4A16 compressed-tensors checkpoint) served text-only on 2× DGX Spark (GB10, sm_121, TP=2) loads, captures CUDA graphs, and serves — but every generated token is token id 0. Output is a run of NUL bytes (`\x00` × max_tokens) on both `/v1/chat/completions` and `/v1/completions`, at temperature 0, for every prompt tried.** No error, no NaN assert, no warning: the engine reports success with a broken forward.

Backend selection logged at startup: `main_attn=triton, msa_decode=False`, with `disable_value_layers=[3..59]` (the value-less sparse layers). So the path exercised is the **Triton MiniMaxSparse attention path on sm_121** over a **compressed-tensors `pack-quantized` int4 (W4A16, g128) checkpoint**. I can't tell from the outside whether the zero logits come from the sparse-attention kernel, the compressed-tensors GEMM path on sm_121, or their combination — but the same weights serve correctly on vLLM on the same two nodes (details below), so the checkpoint itself is sound.

This looks like the same class as #32021 (SM121, MiniMax-M3, NVFP4) and the SM12x "computes wrong rather than crashing" behaviour that the Ling-3.0-flash GB10 cookbook needed a clamp fix for (#36364) — a silent-numerics failure on sm_121 rather than an unsupported-kernel error. A post-load numeric self-check (e.g. refuse to serve if a warm-up generation argmaxes token 0 for N consecutive steps) would have turned this into an actionable error.

**What works on the same setup:**
- Registry resolves the checkpoint: `MiniMaxM3SparseForConditionalGeneration` (VL) is recognised, compressed-tensors quant is picked up, all 59 shards load on sm_121.
- TP=2 over 2 nodes initialises; MiniMaxSparse backend + CUDA graph capture complete (prefill + decode capture ≈ 13 min).
- Server answers requests promptly — decode ≈ 10 tok/s (no speculative decoding) — it is only the *content* that is wrong.

**Control:** the identical safetensors, served by vLLM (a3refaat spark-vllm harness, W4A16 + EAGLE3, TP=2 on the same two Sparks) produce correct, coherent output at ~30+ tok/s and pass our depth probes. So: same weights, same hardware, same driver — SGLang emits zeros, vLLM does not.

### Reproduction

Two-node TP=2 launch (reconstructed from our run notes — flags I'm certain of; I no longer have the shell history of the exact line):

```
# on both nodes (rank 0 / rank 1), image lmsysorg/sglang:v0.5.18-cu130 (arm64 manifest)
docker run --rm --net=host --ipc=host --device nvidia.com/gpu=all \
  -v /mnt/models:/mnt/models -v $HOME/m3-sglang-shadow:/model \
  lmsysorg/sglang:v0.5.18-cu130 \
  python3 -m sglang.launch_server --model-path /model \
    --tp 2 --nnodes 2 --node-rank <0|1> --dist-init-addr 10.100.0.2:5000 \
    --mem-fraction-static 0.945 --context-length 16384 \
    --host 0.0.0.0 --port 8001 --trust-remote-code
```

`/model` is a "shadow" copy of `Sebesky/MiniMax-M3-W4A16-GPTQ` (GPTQ via llm-compressor, compressed-tensors `pack-quantized`, int4 sym g128 Linear / g64 Embedding, lm_head + router gates + indexer projections left bf16; base `MiniMaxAI/MiniMax-M3`) with the safetensors symlinked and **config surgery to serve it text-only**, because neither `--language-model-only` nor `--language-only` applies to this model:
- `text_config` flattened to the top level
- `model_type: "minimax_m2"` (the M3 text stack reuses M2's config class, per `srt/configs/minimax_vl.py`)
- `architectures: ["MiniMaxM3SparseForCausalLM"]`
- `quantization_config` kept verbatim; the checkpoint's custom `*.py` copied; `preprocessor_config.json` / processor configs deleted

Then, at temperature 0:

```
curl -s http://10.100.0.2:8001/v1/chat/completions -H 'Content-Type: application/json' \
  -d '{"model":"/model","temperature":0,"max_tokens":64,"messages":[{"role":"user","content":"In one sentence: what is speculative decoding?"}]}'
```

→ `content` is 64 × `\x00`. Same for `/v1/completions` with a raw prompt. Our probe script (three stages: printable-content check, long-generation depth, ~24K-token prefill) fails at stage 1 every time.

Notes for anyone reproducing on GB10: the weights-only memory floor is `--mem-fraction-static ≈ 0.936` (0.941 with the vision tower), so with 0.945 the KV pool is only ~20K tokens — memory accounting on this model is much fatter than vLLM's (which fits 131K context at 0.92). That is a separate (performance) observation, not the bug.

### Environment

- Hardware: 2 × NVIDIA DGX Spark (GB10, sm_121, 128 GB unified memory each), 200 GbE (ConnectX-7) between nodes, TP=2 (one GPU per node)
- Driver: 580.173.02; CUDA 13.0 toolchain in the image
- Image: `lmsysorg/sglang:v0.5.18-cu130` (arm64), i.e. torch 2.13.0+cu130 and the FlashInfer/sgl-kernel that ship with it; Docker 29.2.1 with CDI GPU access
- Checkpoint: `Sebesky/MiniMax-M3-W4A16-GPTQ` (compressed-tensors, pack-quantized int4 W4A16), 60 layers (3 dense + 57 MoE, 128 experts), text-only via the shadow config above
- Attention path (from server log): `main_attn=triton`, `msa_decode=False`, `disable_value_layers=[3..59]`; CUDA graphs enabled

I can re-run on v0.5.19-cu130 and attach full server logs / a `--log-level debug` startup transcript if that helps localise it, and can test candidate flags (e.g. a different attention/quant backend) on the same pair when a maintenance window is free.

**Comment by IDEA-V** [comment]
Thanks for the thorough report — the config details and the vLLM control were what made this tractable.

I dug into it, and the short version is: this is not sm_121-specific. The cause is in weight loading, it is architecture-independent, and it would reproduce identically on an H100.

The quantized input embedding never loads

Your checkpoint quantizes the embedding table, not just the linears:

`"group_0": {"targets": ["Embedding"], "weights": {"num_bits": 4, "group_size": 64}}`

so it carries embed_tokens.weight_packed / weight_scale / weight_shape and no embed_tokens.weight.

SGLang's compressed-tensors `get_quant_method `handles `LinearBase`, `ParallelLMHead`, `RadixAttention `and `FusedMoE`. A plain `VocabParallelEmbedding` falls through all four, returns `None`, and the layer silently falls back to `UnquantizedEmbeddingMethod` — which builds a bf16 `model.embed_tokens.weight` that nothing in the checkpoint can fill. 

Loading then produces three warnings and carries on:

```
Parameter model.embed_tokens.weight_packed not found in params_dict
Parameter model.embed_tokens.weight_scale  not found in params_dict
Parameter model.embed_tokens.weight_shape  not found in params_dict
```

No exception is raised. The server starts, captures CUDA graphs, and serves a model whose embedding table is still at its allocation-time value. A freshly allocated device buffer comes back zero-filled, so an unloaded embedding table gives zero hidden states, exactly-zero logits, and argmax == 0 at every position. 

vLLM supports exactly this case (CompressedTensorsEmbeddingWNA16Int, vllm-project/vllm#44340 — it registers those same three parameters and does a dequant-gather on lookup), and raises a clear `ValueError` for embedding quantization forms it cannot honour. 

I'm picking up the silent-failure part, which I think is the more important bug here: a model that fails to load its weights should never reach the serving loop. Two PRs:

**PR 1 — compressed-tensors: stop silently ignoring a quantized `Embedding`**

Add the missing `VocabParallelEmbedding` branch to `get_quant_method`, so that a checkpoint declaring embedding quantization SGLang cannot honour raises a clear error instead of falling back to `UnquantizedEmbeddingMethod` and building a parameter that the checkpoint can never fill. This is the direct fix for what you hit, and it mirrors what vLLM already does.

Porting full int4 group-quantized embedding support (`WNA16` in the compressed-tensors schemes) is a separate feature change, and per the module README ("we have only applied the compressed_tensors format of `w8a8_fp8`") it deserves its own issue — happy to open one if there is interest.

**PR 2 — `minimax_m3`: stop reporting unmatched weights as loaded**

A separate bug I ran into while tracing this. In `load_weights`, the `Parameter ... not found in params_dict` branch falls through to `loaded_params.add(name)` because it is missing a `continue` (compare `kimi_k3.py`, which `continue`s on every unmatched path), so the returned coverage set contains names that are not in `params_dict` at all. That return value is what `model_loader/expert_pack_loader.py` already uses for its missing-parameter check, so it needs to be trustworthy. Small fix plus a regression test.

**Comment by michaelmanly** [comment]
@robertforbes68 
nice catch on the silent load failure.

are you doing these sglang/vllm cross-checks pretty often across different quantized models, or was this mostly a one-off for minimax?
